identity-actor/src/actor/actor_builder.rs

```diff
@@ -0,0 +1,274 @@
+    Self {
+      listening_addresses: vec![],
+      keypair: None,
+
+      config: ActorConfig::default(),
+      handlers: HashMap::new(),
+      objects: HashMap::new(),
+    }
+  }
+
+  /// Set the keypair from which the `PeerId` of the actor is derived.
+  ///
+  /// If unset, a new keypair is generated.
+  #[must_use]
+  pub fn keypair(mut self, keypair: Keypair) -> Self {
+    self.keypair = Some(keypair);
+    self
+  }
+
+  /// Add a [`Multiaddr`] to listen on. This can be called multiple times.
+  #[must_use]
+  pub fn listen_on(mut self, address: Multiaddr) -> Self {
+    self.listening_addresses.push(address);
+    self
+  }
+
+  /// Sets the timeout for the underlying libp2p [`RequestResponse`] protocol.
+  #[must_use]
+  pub fn timeout(mut self, timeout: Duration) -> Self {
+    self.config.timeout = timeout;
+    self
+  }
+
+  /// Grants low-level access to the handler map for use in bindings.
+  #[cfg(feature = "primitives")]
+  pub fn handlers(&mut self) -> &mut SyncHandlerMap {
+    &mut self.handlers
+  }
+
+  /// Grants low-level access to the object map for use in bindings.
+  #[cfg(feature = "primitives")]
+  pub fn objects(&mut self) -> &mut ObjectMap {
+    &mut self.objects
+  }
+
+  /// Add a new shared state object and returns a [`ActorHandlerBuilder`] which can be used to
+  /// attach handlers and hooks that operate on that object.
+  pub fn add_state<OBJ>(&mut self, state_object: OBJ) -> ActorHandlerBuilder<OBJ>
+  where
+    OBJ: Clone + Send + Sync + 'static,
+  {
+    let object_id: ObjectId = Uuid::new_v4();
+    self.objects.insert(object_id, Box::new(state_object));
+    ActorHandlerBuilder {
+      object_id,
+      handler_map: &mut self.handlers,
+      _marker_obj: PhantomData,
+    }
+  }
+
+  /// Build the actor with a default transport which supports DNS, TCP and WebSocket capabilities.
+  #[cfg(any(not(target_arch = "wasm32"), target_os = "wasi"))]
+  pub async fn build(self) -> ActorResult<Actor> {
+    let dns_transport =
+      libp2p::dns::TokioDnsConfig::system(libp2p::tcp::TokioTcpConfig::new()).map_err(|err| Error::TransportError {
+        context: "unable to build transport",
+        source: libp2p::TransportError::Other(err),
+      })?;
+
+    let transport = dns_transport
+      .clone()
+      .or_transport(libp2p::websocket::WsConfig::new(dns_transport));
+
+    self.build_with_transport(transport).await
+  }
+
+  /// Build the actor with a custom transport.
+  pub async fn build_with_transport<TRA>(self, transport: TRA) -> ActorResult<Actor>
+  where
+    TRA: Transport + Sized + Clone + Send + Sync + 'static,
+    TRA::Output: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    TRA::Dial: Send + 'static,
+    TRA::Listener: Send + 'static,
+    TRA::ListenerUpgrade: Send + 'static,
+    TRA::Error: Send + Sync,
+  {
+    let executor = Box::new(|fut| {
+      cfg_if::cfg_if! {
+        if #[cfg(any(not(target_arch = "wasm32"), target_os = "wasi"))] {
+          tokio::spawn(fut);
```

Comment:
Question: Would it be possible to make this work for an arbitrary async runtime?

Reply:
At compile-time? Yes, probably. At runtime? Tried it before, but wasn't able to get it working.
We could support `async_std` and `tokio` with feature flags. Based on this flag, we'd also have to use different transports in `libp2p`, so this is not completely trivial. Definitely something to look into for stabilization, but would not focus on it right now.